mangaki/mangaki/utils/recommendations.py

```diff
@@ -49,27 +51,25 @@ def get_reco_algo(request, algo_name='knn', category='all'):
     chrono.save('get rated works')
 
     if algo_name == 'knn':
-        queryset = Rating.objects.filter(work__in=already_rated_works)
-        dataset = Dataset()
-        triplets = list(
-            queryset.values_list('user_id', 'work_id', 'choice'))
-        if request.user.is_anonymous:
-            triplets.extend([
-                (request.user.id, work_id, choice)
-                for work_id, choice in current_user_ratings(request).items()
-            ])
-
-        anonymized = dataset.make_anonymous_data(triplets)
+        try:
+            algo = get_algo_backup(algo_name)
+            dataset = get_dataset_backup(algo_name)
+        except FileNotFoundError:
+            triplets = list(
+                Rating.objects.values_list('user_id', 'work_id', 'choice'))
+            chrono.save('get all %d interesting ratings' % len(triplets))
+            dataset, algo = fit_algo(algo_name, triplets)
+        framed_rated_works = pd.DataFrame(list(current_user_ratings(request).items()), columns = ['work_id', 'choice'])
+        framed_rated_works['rating'] = framed_rated_works['choice'].map(rating_values)
+        ratings_from_user = coo_matrix((framed_rated_works['rating'],(np.zeros(len(framed_rated_works['work_id'])), framed_rated_works['work_id'])), shape = (1, algo.nb_works))
```

Comment:
nb_rated_works = len(framed_rated_works)
    ... [0.] * nb_rated_works

is more readable I think.

(Also, directly `shape=(1, algo.nb_works)`, without spaces around.)